add tensor subclass printing support in fx/graph.py

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164403

it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are tensor subclasses

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
    class GraphModule(torch.nn.Module):
        def forward(self, L_input_batch_inputs_: "DTensor(i64[2, 512][512, 1]cuda:0)", L_self_parameters_weight_: "DTensor(f32[202048, 256][256, 1]cuda:0)"):
```

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

bobrenjc93 committed:
add dtensor printing support in fx/graph.py

it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are dtensors

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "DTensor(dtype=i64, shape=[2, 512], device=cuda:0, mesh=DeviceMesh((dim1=8, dim2=2, dim3=2), device: 'cuda', stride: (4, 2, 1)), placements=(Replicate(), Replicate(), Replicate()))", L_self_parameters_weight_: "DTensor(dtype=f32, shape=[202048, 256], device=cuda:0, mesh=DeviceMesh((dim1=8, dim2=2, dim3=2), device: 'cuda', stride: (4, 2, 1)), placements=(Replicate(), Replicate(), Replicate()))"):

```

[ghstack-poisoned]

bdhirsh commented on torch/fx/graph.py:
```diff
@@ -648,8 +648,21 @@ def emit_node(node: Node):
                     node.meta.get("tensor_meta", node.meta.get("example_value", None)),
                 )
+                # DTensor support
+                if isinstance(meta_val, torch.distributed.tensor.DTensor):
```
adding support for DTensor in tlparse is great, although:

(1)  I feel a bit bad about hardcoding DTensor into `torch.fx.Graph`

(2) it's a shame / might be confusing that we get the nice printing for a graph with DTensor inputs, but not e.g. `Float8Tensor` inputs

bdhirsh replied:
one thing we could do is generaically check to see if the meta_val is a "traceable wrapper subclass", and if so, `__repr__` it. One potential risk of doing this though is that we are calling `repr()` on a subclass that may now hold FakeTensors / SymInts, which the subclass's `repr` may not handle well (what if their repr() calls `item()` on the inner tensor? they shouldn't do this, since it's not "PT2 friendly", but we probably shouldn't crash).

I think my take would be that we still do this generic `repr()` call for any pt2-traceable subclasses, but try/catch it and fall back to a slightly worse print if repr fails. How does that sound?

bobrenjc93 replied:
that's a fair point, let me see if i can refactor this so all tensor subclasses can override a method to get nice tlparse printing

bobrenjc93 committed:
Update on "add tensor subclass printing support in fx/graph.py"


it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are tensor subclasses

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
    class GraphModule(torch.nn.Module):
        def forward(self, L_input_batch_inputs_: "DTensor(i64[2, 512][512, 1]cuda:0)", L_self_parameters_weight_: "DTensor(f32[202048, 256][256, 1]cuda:0)"):
```

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bobrenjc93 commented:
updated the implementation for now to be more generic and check if it's a tensor subclass and if so just wrap the existing tensor descriptor with the subclass name. maybe in the future i'll extend this to have custom printers, but this solves the high level problem for now.

bobrenjc93 committed:
Update on "add tensor subclass printing support in fx/graph.py"


it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are tensor subclasses

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
    class GraphModule(torch.nn.Module):
        def forward(self, L_input_batch_inputs_: "DTensor(i64[2, 512][512, 1]cuda:0)", L_self_parameters_weight_: "DTensor(f32[202048, 256][256, 1]cuda:0)"):
```

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add tensor subclass printing support in fx/graph.py"


it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are tensor subclasses

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
    class GraphModule(torch.nn.Module):
        def forward(self, L_input_batch_inputs_: "DTensor(i64[2, 512][512, 1]cuda:0)", L_self_parameters_weight_: "DTensor(f32[202048, 256][256, 1]cuda:0)"):
```

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add tensor subclass printing support in fx/graph.py"


it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are tensor subclasses

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
    class GraphModule(torch.nn.Module):
        def forward(self, L_input_batch_inputs_: "DTensor(i64[2, 512][512, 1]cuda:0)", L_self_parameters_weight_: "DTensor(f32[202048, 256][256, 1]cuda:0)"):
```

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

ezyang commented on torch/fx/graph.py:
```diff
@@ -648,8 +648,31 @@ def emit_node(node: Node):
                 )
+                # Tensor subclass printing support
+                if (
+                    issubclass(meta_val.__class__, torch.Tensor)
```
This is still not very idiomatic. More idiomatic would be `isinstance(meta_val, torch.Tensor) and type(meta_val) is not torch.Tensor`

ezyang commented on torch/fx/graph.py:
```diff
@@ -648,8 +648,31 @@ def emit_node(node: Node):
+                    )
+                    # Wrap with the tensor subclass name
+                    class_name = meta_val.__class__.__name__
+                    maybe_type_annotation = f': "{class_name}({tensor_repr})"'
```
This is duplicatey with code below

ezyang commented:
I'd like this to land but it needs more work

bdhirsh commented on torch/fx/graph.py:
```diff
@@ -648,8 +648,31 @@ def emit_node(node: Node):
+                    and not isinstance(meta_val, torch._subclasses.FakeTensor)
+                ):
+                    # Get the basic tensor representation
+                    stride_annotation = (
```
if we're just going to print the subclass name and not the other medata, we can probably simplify this. stride/device/shape are all generated from the other codepath too?

bobrenjc93 committed:
Update on "add tensor subclass printing support in fx/graph.py"


it was previously quite misleading since it looks like the inputs to the
dynamo graph are plain tensors when in reality they are tensor subclasses

before
```
class GraphModule(torch.nn.Module):
    def forward(self, L_input_batch_inputs_: "i64[2, 512][512, 1]cuda:0", L_self_parameters_weight_: "f32[202048, 256][256, 1]cuda:0"):
```

after
```
    class GraphModule(torch.nn.Module):
        def forward(self, L_input_batch_inputs_: "DTensor(i64[2, 512][512, 1]cuda:0)", L_self_parameters_weight_: "DTensor(f32[202048, 256][256, 1]cuda:0)"):
```

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
